### PR TITLE
Move array context functionality over from meshmode

### PR DESCRIPTION
**Pull request by inducer**
cc @alexfikl 

FYI I'll merge this right away when it passes, since it's somewhat urgent, what with holding up @kaushikcfd and @matthiasdiener waiting  to continue hacking. :) But please do take a look and let me know if I've screwed up somewhere.

### CODE REVIEW

**Comment by alexfikl** [comment]
Didn't see anything fishy, but it was a very quick skim.

Good to go from my end when the CI passes!

File: arraycontext/container/__init__.py [hunk 1]
```diff
@@ -0,0 +1,220 @@
...
+    """
+    return (
+            cls is ArrayContainer
+            or (serialize_container.dispatch(cls)
+                is not serialize_container.__wrapped__))
+
+
+def is_array_container(ary: Any) -> bool:
+    """Return *True* if the instance *ary* has a registered implementation of
+    :func:`serialize_container`.
```
Comment (alexfikl) [inline arraycontext/container/__init__.py:139]:
```suggestion
    """
    :returns: *True* if the instance *ary* has a registered implementation of
        :func:`serialize_container`.
```

File: arraycontext/impl/pyopencl.py [hunk 2]
```diff
@@ -0,0 +1,381 @@
...
+
+# }}}
+
+
+# {{{ PyOpenCLArrayContext
+
+class PyOpenCLArrayContext(ArrayContext):
+    """
+    A :class:`ArrayContext` that uses :class:`pyopencl.array.Array` instances
+    for DOF arrays.
```
Comment (alexfikl) [inline arraycontext/impl/pyopencl.py:211]:
```suggestion
    for its base array class.
```
?

File: arraycontext/impl/pyopencl.py [hunk 3]
```diff
@@ -0,0 +1,381 @@
...
+        # Work around https://github.com/inducer/pyopencl/pull/402
+        return ary._new_with_changes(
+                data=None, offset=0, shape=(0,), strides=(ary.dtype.itemsize,))
+    if ary.flags.f_contiguous:
+        return ary.reshape(-1, order="F")
+    elif ary.flags.c_contiguous:
+        return ary.reshape(-1, order="C")
+    else:
+        raise ValueError("cannot flatten group array of DOFArray for norm, "
+                f"with strides {ary.strides} of {ary.dtype}")
```
Comment (alexfikl) [inline arraycontext/impl/pyopencl.py:151]:
```suggestion
        raise ValueError("cannot flatten array "
                f"with strides {ary.strides} of {ary.dtype}")
```

**Review by alexfikl — APPROVED** [review]
Found some nitpicks!

**Comment by inducer** [comment]
Thanks for taking a look!

**Comment by inducer** [comment]
:tada:

**Pull request merged**